### PR TITLE
Convert PaginatedResource to TypeScript

### PR DESCRIPTION
**Pull request by owenpearson**

### CODE REVIEW

File: common/lib/client/paginatedresource.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+import { PaginatedResultCallback } from '../../types/utils';
+
+// TODO: Replace this once rest.js is converted to TypeScript
+type Rest = any;
+
+export type BodyHandler = (body: unknown, headers: Record<string, string>, packed?: boolean) => any;
+
+function getRelParams(linkUrl: string) {
+	const urlMatch = linkUrl.match(/^\.\/(\w+)\?(.*)$/);
+	return urlMatch && Utils.parseQueryString(urlMatch[2]);
```
Comment (ikbalkaya) [inline common/lib/client/paginatedresource.ts:14]: Should you be also checking `urlMatch.size > 2` ?

File: common/lib/client/paginatedresource.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	const relParams: Record<string, Record<string, string>> = {};
+	for(let i = 0; i < linkHeader.length; i++) {
+		const linkMatch = linkHeader[i].match(/^\s*<(.+)>;\s*rel="(\w+)"$/);
+		if(linkMatch) {
+			const params = getRelParams(linkMatch[1]);
+			if(params)
+				relParams[linkMatch[2]] = params;
+		}
+	}
+	return relParams;
```
Comment (ikbalkaya) [inline common/lib/client/paginatedresource.ts:30]: There are a couple of unsafe array accesses on this function. I am not sure if those functions have their corresponding unit tests but I think it would be a great way to discover and fix those edge cases

File: common/lib/client/paginatedresource.ts [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+			Logger.logAction(Logger.LOG_ERROR, 'PaginatedResource.handlePage()', 'Unexpected error getting resource: err = ' + Utils.inspectError(err));
+			callback(err);
+			return;
+		}
+		let items, linkHeader, relParams;
+		try {
+			items = this.bodyHandler(body, headers, unpacked);
+		} catch(e) {
+			/* If we got an error, the failure to parse the body is almost certainly
+			 * due to that, so cb with that in preference to the parse error */
```
Comment (ikbalkaya) [inline common/lib/client/paginatedresource.ts:105]: "so cb with that in preference to the parse error" I think there is a typo here

File: common/lib/client/paginatedresource.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+	hasNext?: () => boolean;
+	isLast?: () => boolean;
+
+	constructor(resource: PaginatedResource, items: T[], relParams?: Record<string, any>) {
+		this.resource = resource;
+		this.items = items;
+
+		if(relParams) {
+			if('first' in relParams) {
+				this.first = (cb: (result?: ErrorInfo | null) => void) => {
```
Comment (ikbalkaya) [inline common/lib/client/paginatedresource.ts:137]: Ah. Now I can get what 'cb' is. Can this be named `callback`?

**Review by ikbalkaya — DISMISSED** [review]
Looks good overall. However there is a pattern I recognise where arrays or parsed content is used without safety check. I'm not sure if there is a plan in pipeline but I think writing some unit tests that covers all cases would make those a bit more obvious.

File: common/lib/client/paginatedresource.ts (hunk 4 above)
Comment (QuintinWillison) [reply common/lib/client/paginatedresource.ts:137]: I can't disagree with this. 😄 Though, perhaps, worth acknowledging that this variable name was inherited from the original JS implementation.

**Review by QuintinWillison — APPROVED** [review]

File: common/lib/client/paginatedresource.ts (hunk 1 above)
Comment (owenpearson) [reply common/lib/client/paginatedresource.ts]: 2e1fd3b1

File: common/lib/client/paginatedresource.ts (hunk 3 above)
Comment (owenpearson) [reply common/lib/client/paginatedresource.ts]: cb is short for callback but I agree that it's still not worded great so I improved it in cdfa6f7

File: common/lib/client/paginatedresource.ts (hunk 4 above)
Comment (owenpearson) [reply common/lib/client/paginatedresource.ts]: 500ea75

File: common/lib/client/paginatedresource.ts (hunk 2 above)
Comment (owenpearson) [reply common/lib/client/paginatedresource.ts:30]: You're right, but I don't want to change this now. It's always worked up until this point so I would assume it's safe, even if it would take a lot of work to prove it.

File: common/lib/client/paginatedresource.ts (hunk 1 above)
Comment (QuintinWillison) [reply common/lib/client/paginatedresource.ts]: Resolving this conversation on @ikbalkaya's behalf.

File: common/lib/client/paginatedresource.ts (hunk 2 above)
Comment (QuintinWillison) [reply common/lib/client/paginatedresource.ts:30]: Resolving this conversation on @ikbalkaya's behalf.

File: common/lib/client/paginatedresource.ts (hunk 3 above)
Comment (QuintinWillison) [reply common/lib/client/paginatedresource.ts]: Resolving this conversation on @ikbalkaya's behalf.

File: common/lib/client/paginatedresource.ts (hunk 4 above)
Comment (QuintinWillison) [reply common/lib/client/paginatedresource.ts]: Resolving this conversation on @ikbalkaya's behalf.
